source/agora/consensus/EnrollmentManager.d

```diff
@@ -532,6 +532,71 @@ public class EnrollmentManager
         return false;
     }
 
+    /***************************************************************************
+
+        Get all the current validators
+
+        Params:
+            validators = will be filled with all the validators during
+                their validation cycles
+
+        Returns:
+            Return true if there is no error in getting validators
+
+    ***************************************************************************/
+
+    public bool getValidators (out Enrollment[] validators) @safe nothrow
+    {
+        try
+        {
+            () @trusted {
+                auto results = this.db.execute("SELECT val FROM validator_set" ~
+                    " WHERE enrolled_height is not null");
+                foreach (row; results)
+                {
+                    auto enroll =
+                        deserializeFull!Enrollment(row.peek!(ubyte[])(0));
+                    validators ~= enroll;
+                }
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error, exception:{}", ex);
+            return false;
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Set current block height
+
+        The enrollment manager can weed out expired validators from the set
+        based on the block height.
+
+        Params:
+            block_height = current block height
+
+    ***************************************************************************/
+
+    public void setCurrentBlockHeight (ulong block_height) @safe nothrow
+    {
+        try
+        {
+            () @trusted {
+                this.db.execute("DELETE FROM validator_set " ~
+                    "WHERE enrolled_height <= ?", block_height - ValidatorCycle);
```

Comment:
What happens when we call this with `block_height == 1` ?

Reply:
> What happens when we call this with block_height == 1 ?

There seems no problem but I will add unit test code for that and check it.

Reply:
Really ? It substract a large unsigned value to a smaller unsigned value.
So either we get negative number, or we wrap-around and get a very high number.
In any case, not something that should happen.

Reply:
It passed new unit test code as follows.
```
man.setCurrentBlockHeight(1);
assert(man.getValidators(validators));
assert(validators.length == 2);
```
I guess the sqlite get the very big number as small singed integer by the operation of the two's complement. But for the code to be safer, I will add the code checking the block_height.